main.js

```diff
@@ -134,13 +147,39 @@ function extractData() {
 	};
 }
 
+function getCleanGradesFromExams(exams) {
+	var clean = []
+	var dirty_exams = exams.get()
+	var found = 0;
+	for(var i = 0; i < dirty_exams.length; i++) {
+		for(var j = 0; j < clean.length; j++) {
+			if(clean[j].course_id == dirty_exams[i].course_id && dirty_exams[i].grade != '') {
+				clean[j] = dirty_exams[i];
+				found = 1;
+			}
+		}
+		if(!found && dirty_exams[i].grade != '')
+			clean.push(dirty_exams[i]);
+		found = 0;
+	}
+	var rclean = clean.reverse();
+	return rclean;
+}
+
+
 function rearrangeDocument(exams, totals, details, summary) {
 	const original = $('body > div').html();
 	console.log(original);
 	withFile("ug.html", function(data) {
 		$('html').html(data); 
 		
-		for (const e of exams)		
+		
+		var clean = getCleanGradesFromExams(exams);
+		for (const c of clean) 
```

Comment:
Consider inlining the variable `clean` (a matter of taste, but I think it's better).